src/publication.rs

```diff
@@ -5,6 +5,47 @@ use crate::{
     ProtocolError, QoS, Retain,
 };
 
+pub trait ToPayload {
```

Comment:
I don't get why it needs to be `Fn` (or `FnMut`) and not `FnOnce`.

And what about this?

```rust
enum Payload<'a, F: FnOnce(&mut [u8]) -> Result<usize, ()>> {
    Borrowed(&'a [u8]),
    Callback(F),
}
```

Reply:
You can also just have the callback and then convert `Borrowed(src) => Callback(|buf| buf[..src.len()].copy_from_slice(src))`

Reply:
The problem with the enum is that it doesn't have a concrete type in the case where we're only using a `&[u8]` slice. We need the contrete type for `ReceivedPacket`, so there's a catch-22 there. I originally started iwth an enum and then ran into this.

Reply:
Maybe split `Pub` packet into `IncomingPub` and `OutgoingPub`?

The idea of a callback for payload is pretty ambitious anyway. It requires fixing many additional variable length things (like the topic to respond to or properties) before knowing whether the payload generation/serialization even succeeds. We'd want to invert the serialization. We already do this for the fixed header with the remaining length varint. Three years ago we had the full reverse serialization (building the packet from the payload backwards).

Reply:
For example:
```
error[E0283]: type annotations needed
   --> src\mqtt_client.rs:480:40
    |
480 |                 let Ok(response_pub) = minimq::Publication::new(response.msg.as_bytes())
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `F` declared on the struct `Publication`
...
490 |                 if client.publish(response_pub).is_err() {
    |                           ------- type must be known at this point
    |
    = note: multiple `impl`s satisfying `for<'a> _: FnOnce<(&'a mut [u8],)>` found in the following crates: `alloc`, `core`:
            - impl<A, F> FnOnce<A> for &F
              where A: Tuple, F: Fn<A>, F: ?Sized;
            - impl<A, F> FnOnce<A> for &mut F
              where A: Tuple, F: FnMut<A>, F: ?Sized;
            - impl<Args, F, A> FnOnce<Args> for std::boxed::Box<F, A>
              where Args: Tuple, F: FnOnce<Args>, A: Allocator, F: ?Sized;
note: required by a bound in `minimq::packets::OutgoingPub`
   --> C:\Users\rsummers\Documents\repositories\quartiq\minimq\src\packets.rs:77:34
    |
77  | pub struct OutgoingPub<'a, E, F: FnOnce(&mut [u8]) -> Result<usize, E>> {
    |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `OutgoingPub`
help: consider specifying the generic arguments
    |
480 |                 let Ok(response_pub) = minimq::Publication::<E, F>(response.msg.as_bytes())

```